arangod/Replication2/ReplicatedLog/LogFollower.cpp

```diff
@@ -0,0 +1,413 @@
+  return _guardedFollowerData.doUnderLock([this](GuardedFollowerData& followerData) {
+    LOG_CTX("838fe", DEBUG, _loggerContext) << "follower resign";
+    if (followerData._logCore == nullptr) {
+      LOG_CTX("55a1d", WARN, _loggerContext)
+          << "follower log core is already gone. Resign was called twice!";
+      ASSERT_OR_THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_FOLLOWER_RESIGNED);
+    }
+
+    // use a unique ptr because move constructor for multimaps is not noexcept
+    auto queue = std::make_unique<WaitForQueue>(
+        std::move(followerData._waitForQueue.getLockedGuard().get()));
+
+    auto action = [queue = std::move(queue)]() mutable noexcept {
+      std::for_each(queue->begin(), queue->end(), [](auto& pair) {
+        pair.second.setException(basics::Exception(TRI_ERROR_REPLICATION_LEADER_CHANGE,
+                                                   __FILE__, __LINE__));
+      });
+    };
+    using action_type = decltype(action);
+
+    static_assert(std::is_nothrow_move_constructible_v<action_type>);
+    static_assert(std::is_nothrow_constructible_v<DeferredAction, std::add_rvalue_reference_t<action_type>>);
+
+    // make_tuple is noexcept, _logCore is a unique_ptr which is nothrow move constructable
+    return std::make_tuple(std::move(followerData._logCore),
+                           DeferredAction{std::move(action)});
+  });
+}
+
+replicated_log::LogFollower::LogFollower(LoggerContext const& logContext,
+                                         std::shared_ptr<ReplicatedLogMetrics> logMetrics,
+                                         ParticipantId id, std::unique_ptr<LogCore> logCore,
+                                         LogTerm term, std::optional<ParticipantId> leaderId,
+                                         replicated_log::InMemoryLog inMemoryLog)
+    : _logMetrics(std::move(logMetrics)),
+      _loggerContext(
+          logContext.with<logContextKeyLogComponent>("follower")
+              .with<logContextKeyLeaderId>(leaderId.value_or("<none>"))
+              .with<logContextKeyTerm>(term)),
+      _participantId(std::move(id)),
+      _leaderId(std::move(leaderId)),
+      _currentTerm(term),
+      _guardedFollowerData(*this, std::move(logCore), std::move(inMemoryLog)) {
+  _logMetrics->replicatedLogFollowerNumber->fetch_add(1);
+}
+
+auto replicated_log::LogFollower::waitFor(LogIndex idx)
+    -> replicated_log::ILogParticipant::WaitForFuture {
+  auto self = _guardedFollowerData.getLockedGuard();
+  if (self->_commitIndex >= idx) {
+    return futures::Future<std::shared_ptr<QuorumData const>>{
+        std::in_place, std::make_shared<QuorumData>(idx, _currentTerm)};
+  }
+  // emplace might throw a std::bad_alloc but the remainder is noexcept
+  // so either you inserted it and or nothing happens
+  auto it = self->_waitForQueue.getLockedGuard()->emplace(idx, WaitForPromise{});
+  auto& promise = it->second;
+  auto&& future = promise.getFuture();
```

Comment:
Is this guaranteed to be race-free?
The emplace is executed under the lock, but dereferencing the iterator later IMO is executed without the lock. Is it possible that the item gets removed from the queue by another thread in-between?

Reply:
To be honest I don't know why the `_waitForQueue` is again wrapped in a `Guarded` structure. By the code design you should only ever have access to this member if you hold the lock, which here is guarded by `self` and held until the function ends.

I will double check with @goedderz but I think this has historical reasons.

Reply:
Looks like it on the first glance, but it's not. It was introduced in https://github.com/arangodb/arangodb/pull/14486/commits/945fd82680ad0c535e991d7f923096812ce9ba4a. It is there because in `appendEntries`, the `waitForQueue` is resolved *after* `_guardedFollowerData` is unlocked.

Now I think this is race-free nonetheless, because the `_waitForQueue`-lock is not (and cannot be) acquired without already holding `_guardedFollowerData`-lock.

We should at least add a comment at the declaration of `Guarded<WaitForQueue> _waitForQueue;` in `GuardedFollowerData` why this guard is needed. And I propose that we dereference the iterator under the `_waitForQueue`-lock: It is more obviously correct and thus either to read, and it is not a pessimization, because every concurrent access has to hold the `_guardedFollowerData`-lock anyway. And/or we add an additional comment.